### PR TITLE
Refactor web service from Python to Ruby

### LINKED ISSUE #11: [S02] Refactor minitwit.py

### PR DESCRIPTION
**Pull request by janschill**
This refactor is porting the web service MiniTwit from a Python implementation to Ruby using frameworks Roda and Sequel.

[Roda](http://roda.jeremyevans.net/index.html) is a lightweight, fast, scalable framework to build web services. It is implementing a routing tree, which makes its routing faster than say Sinatra's list implementation. Roda comes also with the possibility to include plugins that allow for example simple session implementations.
On top of Roda this PR also uses the [Sequel](http://sequel.jeremyevans.net/) framework as an SQL abstraction layer and the mapping of app models to database tables. This brings the benefit of being able to switch databases easily by just changing the connector, because no SQL is written anywhere. Also the mapping of models allow an intuitive creation of new instances of models/rows.

Rubocop is a Ruby gem that lints Ruby code. When run with `rubocop` in the project root, it will highlight all the issues that are currently present in the code. It can be installed with Bundler from the Gemfile. There is also an official VS Code support with the [Ruby extension](https://marketplace.visualstudio.com/items?itemName=rebornix.Ruby), which will dynamically highlight bad code and fix as much as possible on format.

This will resolve #11 

Some things are still missing:
- [x] CSS file is not loading
- [x] Follow and unfollow of users
- [ ] Retrieve own messages and the ones from followers
- [ ] flag_tool (done in different PR)
- [ ] control.sh (done in different PR)
- [ ] Date conversion of messages in frontend

### CODE REVIEW

**Comment by alevani** [comment]
Could you please explicitly say what are the files we should look at and the one you made (aka not generated or remaining file from other features) thanks :+1:

**Comment by janschill** [comment]
> Could you please explicitly say what are the files we  should look at and the one you made (aka not generated or remaining file  from other features) thanks

Good start point would be the `config.ru` file from there it goes into the `minitwit.rb` and then just follow the `require` etc.
The `Gemfile` is a generated file that specifies the project dependencies (gems).

File: release-02/code/flag_tool.rb [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    end
+    exit(true)
+end
+
+if ARGV.length >= 1 && !(ARGV[0].eql? "-i") && !(ARGV[0].eql? "-h") then
+    for id in ARGV
+        messages.where(message_id: id).update(:flagged => 1);
+    end
+    exit(true)
+end
```
Comment (alevani) [inline release-02/code/flag_tool.rb]: I don't understand why this file is part of the pull request if it is about the refactoring of the web framework. If you want to send it to submission, I think you should open a new PR as it is important enough to be a "feature"

File: release-02/refactor/bin/control.rb [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+  puts 'Starting MiniTwit...'
+elsif ARGV[0] == 'stop'
+  puts 'Stopping MiniTwit...'
+elsif ARGV[0] == 'inspectdb'
+
+elsif ARGV[0] == 'flag'
+
+else
+  puts 'I do not know this command...'
+end
```
Comment (alevani) [inline release-02/refactor/bin/control.rb]: Same thing for this file, it seems that this is not even completed, so this should not be part of a PR to master. As proposed above, you should complete it (if you want) and do a new PR to master.

**Comment by alevani** [comment]
General comment : I think the name of the PR is not explicit enough, when I look at "ruby refactor" I cannot say what the PR is about, and in the description I think you should say something like "this is the refactor form python to ruby of the web app" 😃

File: release-02/code/flag_tool.rb (hunk 1 above)
Comment (janschill) [reply release-02/code/flag_tool.rb]: Yes, good catch. I accidentally put it in there. Will remove it.

File: release-02/refactor/bin/control.rb (hunk 2 above)
Comment (janschill) [reply release-02/refactor/bin/control.rb]: File is removed.

**Comment by janschill** [comment]
> General comment : I think the name of the PR is not explicit enough, when I look at "ruby refactor" I cannot say what the PR is about, and in the description I think you should say something like "this is the refactor form python to ruby of the web app" 😃

Done.

File: release-02/refactor/Gemfile [hunk 3]
```diff
@@ -0,0 +1,11 @@
+# frozen_string_literal: true
+
+source "https://rubygems.org"
+
+git_source(:github) {|repo_name| "https://github.com/#{repo_name}" }
+
+gem 'roda'
```
Comment (Fr3d-) [inline release-02/refactor/Gemfile]: I think we have to add `sqlite3` as a dependency, I got errors running `rackup` before I installed it.

Comment (janschill) [reply release-02/refactor/Gemfile]: You are absolutely right. I wonder why it is not in there.

Comment (janschill) [reply release-02/refactor/Gemfile]: Has been added.

**Pull request merged**